core/kernel/tee_ta_manager.c

```diff
@@ -821,79 +824,16 @@ bool tee_ta_session_is_cancelled(struct tee_ta_session *s, TEE_Time *curr_time)
 	return false;
 }
 
-static void update_current_ctx(struct thread_specific_data *tsd)
-{
-	struct tee_ta_ctx *ctx = NULL;
-	struct tee_ta_session *s = TAILQ_FIRST(&tsd->sess_stack);
-
-	if (s) {
-		if (is_pseudo_ta_ctx(s->ctx))
-			s = TAILQ_NEXT(s, link_tsd);
-
-		if (s)
-			ctx = s->ctx;
-	}
-
-	if (tsd->ctx != ctx)
-		tee_mmu_set_ctx(ctx);
-	/*
-	 * If current context is of user mode, then it has to be active too.
-	 */
-	if (is_user_mode_ctx(ctx) != core_mmu_user_mapping_is_active())
-		panic("unexpected active mapping");
-}
-
-void tee_ta_push_current_session(struct tee_ta_session *sess)
-{
-	struct thread_specific_data *tsd = thread_get_tsd();
-
-	TAILQ_INSERT_HEAD(&tsd->sess_stack, sess, link_tsd);
-	update_current_ctx(tsd);
-}
-
-struct tee_ta_session *tee_ta_pop_current_session(void)
-{
-	struct thread_specific_data *tsd = thread_get_tsd();
-	struct tee_ta_session *s = TAILQ_FIRST(&tsd->sess_stack);
-
-	if (s) {
-		TAILQ_REMOVE(&tsd->sess_stack, s, link_tsd);
-		update_current_ctx(tsd);
-	}
-	return s;
-}
-
-TEE_Result tee_ta_get_current_session(struct tee_ta_session **sess)
-{
-	struct tee_ta_session *s = TAILQ_FIRST(&thread_get_tsd()->sess_stack);
-
-	if (!s)
-		return TEE_ERROR_BAD_STATE;
-	*sess = s;
-	return TEE_SUCCESS;
-}
-
-struct tee_ta_session *tee_ta_get_calling_session(void)
-{
-	struct tee_ta_session *s = TAILQ_FIRST(&thread_get_tsd()->sess_stack);
-
-	if (s)
-		s = TAILQ_NEXT(s, link_tsd);
-	return s;
-}
-
 #if defined(CFG_TA_GPROF_SUPPORT)
 void tee_ta_gprof_sample_pc(vaddr_t pc)
 {
-	struct tee_ta_session *s = NULL;
+	struct ts_session *s = ts_get_current_session();
 	struct user_ta_ctx *utc = NULL;
 	struct sample_buf *sbuf = NULL;
 	TEE_Result res = 0;
 	size_t idx = 0;
 
-	if (tee_ta_get_current_session(&s) != TEE_SUCCESS)
-		return;
-	sbuf = s->sbuf;
+	sbuf = to_ta_session(s)->sbuf;
```

Comment:
Previous implementation safely returned without segfault when no current session.
Looks ok here, but maybe cross check this is safe.